### PR TITLE
feat: add the obs gosdk and Implement the CreateVolume interface.

### PR DESCRIPTION
**Pull request by leleyi**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/obs/constants.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const (
+	// PvcNameTag in annotations
+	PvcNameTag = "csi.storage.k8s.io/pvc/name"
+	// PvcNsTag in annotations
+	PvcNsTag = "csi.storage.k8s.io/pvc/namespace"
+	// PvNameKey key
+	PvNameKey = "csi.storage.k8s.io/pv/name"
+
+	// CsiClusterNodeIDKey in volume metadata
+	CsiClusterNodeIDKey = "evs.csi.huaweicloud.com/nodeId"
```
Comment (chengxiangdong) [inline pkg/obs/constants.go:28]: do we need this `CsiClusterNodeIDKey` in OBS

File: pkg/obs/constants.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (chengxiangdong) [inline pkg/obs/constants.go:1]: I think we do not need these constants

File: pkg/obs/controllerserver.go [hunk 3]
```diff
@@ -29,7 +32,51 @@ type controllerServer struct {
...
+	log.Infof("CreateVolume: called with args %v", protosanitizer.StripSecrets(*req))
+	credentials := cs.Driver.cloud
+
+	volName := req.GetName()
+	if err := createVolumeValidation(volName, req.GetVolumeCapabilities()); err != nil {
+		return nil, err
+	}
+
+	if vol, err := services.GetBucket(credentials, volName); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:44]: s/err.Error()/err

File: pkg/obs/controllerserver.go [hunk 4]
```diff
@@ -29,7 +32,51 @@ type controllerServer struct {
...
+	} else if vol != "" {
+		return buildCreateVolumeResponse(volName), nil
+	}
+
+	_, err := services.CreateBucketWithTag(credentials, volName)
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:49]: It is recommended that `CreateBucketWithTag` be split into `CreateBucket` and `CreateBucketTag`.

File: pkg/obs/services/buckets.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+func GetBucketMetadata(c *config.CloudCredentials, bucketName string) (*obs.GetBucketMetadataOutput, error) {
+	client, err := getObsClient(c)
+	if err != nil {
+		return nil, err
+	}
+	input := &obs.GetBucketMetadataInput{}
+	input.Bucket = bucketName
+	output, err := client.GetBucketMetadata(input)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Error, OBS instance %s does not exist, error: %v", bucketName, err)
```
Comment (deer-hang) [inline pkg/obs/services/buckets.go:55]: when err is not nil, if it is necessary to parse err message to verify it is 404?
for other err message, response code should be codes.Internal

File: pkg/obs/controllerserver.go [hunk 6]
```diff
@@ -29,7 +32,51 @@ type controllerServer struct {
...
+	if len(capabilities) == 0 {
+		return status.Error(codes.InvalidArgument, "Validation failed, volume capabilities cannot be empty")
+	}
+	return nil
+}
+
+func buildCreateVolumeResponse(volName string) *csi.CreateVolumeResponse {
+	response := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId: volName,
```
Comment (chengxiangdong) [inline pkg/obs/controllerserver.go:76]: pls append `AccessibleTopology` and `CapacityBytes`

File: pkg/obs/controllerserver.go (hunk 4 above)
Comment (deer-hang) [inline pkg/obs/controllerserver.go:49]: if _, err := services.CreateBucketWithTag(credentials, volName); err != nil {}

File: pkg/obs/controllerserver.go [hunk 7]
```diff
@@ -29,7 +32,51 @@ type controllerServer struct {
...
+		return buildCreateVolumeResponse(volName), nil
+	}
+
+	_, err := services.CreateBucketWithTag(credentials, volName)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = services.GetBucket(credentials, volName)
+	if err != nil {
```
Comment (deer-hang) [inline pkg/obs/controllerserver.go:55]: ditto

File: pkg/obs/services/buckets.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	return output, nil
+}
+
+func CheckBucketExists(c *config.CloudCredentials, bucketName string) (bool, error) {
+	client, err := getObsClient(c)
+	if err != nil {
+		return false, err
+	}
+	_, err = client.HeadBucket(bucketName)
```
Comment (deer-hang) [inline pkg/obs/services/buckets.go:65]: ditto

File: pkg/obs/services/buckets.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	if err != nil {
+		return "", err
+	}
+	input := &obs.CreateBucketInput{}
+	input.Bucket = bucketName
+	input.Location = c.Global.Region
+	input.ACL = obs.AclPrivate
+	input.IsFSFileInterface = true
+	_, err = client.CreateBucket(input)
+	if err == nil {
```
Comment (deer-hang) [inline pkg/obs/services/buckets.go:96]: ditto

File: pkg/obs/services/buckets.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+import (
+	"fmt"
+	"github.com/huaweicloud/huaweicloud-csi-driver/pkg/config"
+	"github.com/huaweicloud/huaweicloud-sdk-go-obs/obs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"net/http"
+)
+
+func GetBucket(c *config.CloudCredentials, bucketName string) (string, error) {
```
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:28]: The `GetBucket`  parameters containts `bucketName` and it return `bucketName`. 
What is this function for?  and the name is confusing.

File: pkg/obs/services/buckets.go [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	return bucketName, nil
+}
+
+func GetBucketMetadata(c *config.CloudCredentials, bucketName string) (*obs.GetBucketMetadataOutput, error) {
+	client, err := getObsClient(c)
+	if err != nil {
+		return nil, err
+	}
+	input := &obs.GetBucketMetadataInput{}
```
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:51]:
```
input := &obs.GetBucketMetadataInput{}
input.Bucket = bucketName
```
change to following is more clear
```
input := &obs.GetBucketMetadataInput{Bucket: bucketName}
```

File: pkg/obs/services/buckets.go (hunk 5 above)
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:55]: Are you sure the `err` just didn't find the bucket?

File: pkg/obs/services/buckets.go [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+func GetBucketCSITagging(c *config.CloudCredentials, bucketName string) (string, error) {
+	client, err := getObsClient(c)
+	if err != nil {
+		return "", err
+	}
+	output, err := client.GetBucketTagging(bucketName)
+	if err != nil {
+		return "", status.Errorf(codes.Internal, "Error, getting OBS instance %s tag, error: %v", bucketName, err)
+	}
+	tag := output.Tags[0].Value
```
Comment (deer-hang) [inline pkg/obs/services/buckets.go:155]: maybe validate len(output.Tags) > 0?

File: pkg/obs/services/buckets.go [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	if obsError, ok := err.(obs.ObsError); ok {
+		if obsError.StatusCode == http.StatusNotFound {
+			return false, nil
+		}
+	}
+	return false, status.Errorf(codes.Internal, "Error, heading OBS instance %s, error: %v", bucketName, err)
+}
+
+func CheckBucketFSStatus(c *config.CloudCredentials, bucketName string) (bool, error) {
```
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:77]: The func name is not clear,  how about `IsParallelFSBucket`

File: pkg/obs/services/buckets.go [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+	}
+	return metadata.FSStatus == obs.FSStatusEnabled, nil
+}
+
+func CreateBucket(c *config.CloudCredentials, bucketName string) (string, error) {
+	client, err := getObsClient(c)
+	if err != nil {
+		return "", err
+	}
+	input := &obs.CreateBucketInput{}
```
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:90]:
```
input := &obs.CreateBucketInput{}
input.Bucket = bucketName
input.Location = c.Global.Region
input.ACL = obs.AclPrivate
input.IsFSFileInterface = true
```
suggest change to following
```
input := &obs.CreateBucketInput{
    Bucket: bucketName,
    Location: c.Global.Region,
    ACL: obs.AclPrivate,
    IsFSFileInterface: true,
}
```

File: pkg/obs/controllerserver.go (hunk 3 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go:44]: Ok， Done

File: pkg/obs/services/buckets.go (hunk 14 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:90]: done

File: pkg/obs/services/buckets.go (hunk 13 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:77]: down

File: pkg/obs/services/buckets.go (hunk 12 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:155]: down

File: pkg/obs/services/buckets.go (hunk 5 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:55]: done

File: pkg/obs/services/buckets.go (hunk 11 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:51]: done

File: pkg/obs/services/buckets.go (hunk 10 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:28]: down

File: pkg/obs/services/buckets.go (hunk 5 above)
Comment (leleyi) [reply pkg/obs/services/buckets.go:55]: down

File: pkg/obs/controllerserver.go (hunk 4 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go:49]: down

File: pkg/obs/constants.go (hunk 2 above)
Comment (leleyi) [reply pkg/obs/constants.go:1]: done

File: pkg/obs/constants.go (hunk 1 above)
Comment (leleyi) [reply pkg/obs/constants.go:28]: done

File: pkg/obs/controllerserver.go (hunk 6 above)
Comment (leleyi) [reply pkg/obs/controllerserver.go]: done

File: pkg/obs/constants.go [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
```
Comment (chengxiangdong) [inline pkg/obs/constants.go:1]: may be we can remove this file

File: pkg/obs/services/objects.go [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+	if err == nil {
+		return output, nil
+	}
+	if obsError, ok := err.(obs.ObsError); ok && obsError.StatusCode == http.StatusNotFound {
+		return nil, status.Errorf(codes.NotFound, "Error OBS instance %s does not exist: %v", bucketName, err)
+	}
+	return output, status.Errorf(codes.Internal, "Error get OBS instance %s upload list: %v", bucketName, err)
+}
+
+// err 1. 桶是否存在， 2. 资源存在与否， 3. 没有的时候如何返回。
```
Comment (chengxiangdong) [inline pkg/obs/services/objects.go:126]: pls remove these comments.

File: pkg/obs/services/objects.go [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+// get 资源不存在404 error中包含， 根据当前方法判断error信息。
+// list 一般不关注资源在不在。
```
Comment (chengxiangdong) [inline pkg/obs/services/objects.go:128]: pls remove these comments

File: pkg/obs/services/buckets.go [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+	if initLog() != nil {
+		return nil, status.Errorf(codes.Internal, "Error initializing OBS client log: %v", err)
+	}
+	return client, nil
+}
+
+func initLog() error {
+	var logFullPath string = "./logs/OBS-SDK.log"
+	var maxLogSize int64 = 1024 * 1024 * 10
+	err := obs.InitLog(logFullPath, maxLogSize, 10, obs.LEVEL_INFO, true)
```
Comment (chengxiangdong) [inline pkg/obs/services/buckets.go:217]: we do not need print log to files, pls disable it.

**Comment by chengxiangdong** [comment]
well done
/lgtm

cc: @niuzhenguo

**Pull request merged**